mitodl/unified-ecommerce
# Integrated System - discount API

### What are the relevant tickets?
https://github.com/mitodl/hq/issues/5892

### Description (What does it do?)
Adds an API endpoint that will be used by integrated systems to create new Discount objects.  This newly added API will only receive POST requests that have an authorization header with a value in the format of "Api-Key <INTEGRATED_SYSTEM_API_KEY>".  The IntegratedSystemAPIKey is a model added in this PR that links an API key with an Integrated System record using a Foreign Key relationship.  When a Discount is created via the API endpoint added in this PR, the Discount will automatically be associated to the IntegratedSystem record that is associated with the IntegratedSystemAPIKey.

### How can this be tested?

1. Run migrations
2. Go to http://ue.odl.local:9080/admin/system_meta/integratedsystemapikey/ and create a new API key that is associated with an IntegratedSystem record.  Save the API key shown in the Django Admin banner after successfully saving the new IntegratedSystemAPIKey.
3. I used Postman for the following steps
4. Create a POST request to http://ue.odl.local:9080/api/v0/payments/discounts/
5. Add an authorization header of: Api-Key <INTEGRATED_SYSTEM_API_KEY_FROM_STEP_1>
6. Add a body to the post request as shown in the screen shot below.
7. Execute the POST request.  Verify that the Discount record was created at http://ue.odl.local:9080/admin/payments/discount/.

<img width="822" alt="Screenshot 2024-12-03 at 9 43 55 AM" src="https://github.com/user-attachments/assets/1c70c13d-d989-4f38-a537-3c39b1a4e3d0">

<img width="818" alt="Screenshot 2024-12-03 at 9 43 49 AM" src="https://github.com/user-attachments/assets/cc648dda-519f-4d5d-aba8-db2213caad07">

Example post request body:
```
{
    "discount_type": "dollars-off",
    "redemption_type": "unlimited",
    "payment_type": "staff",
    "amount": "4",
    "count": 1,
    "codes": ["abc4"],
    "product": "1",
    "users": "1"
}
```
